libnavigator/src/test/java/com/mapbox/navigation/navigator/internal/router/RouterInterfaceDelegateTest.kt

```diff
@@ -0,0 +1,264 @@
+package com.mapbox.navigation.navigator.internal.router
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.bindgen.Expected
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.route.RouteRefreshCallback
+import com.mapbox.navigation.base.route.RouteRefreshError
+import com.mapbox.navigation.base.route.Router
+import com.mapbox.navigation.base.route.RouterCallback
+import com.mapbox.navigation.base.route.RouterFailure
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigator.RouterError
+import com.mapbox.navigator.RouterErrorType
+import com.mapbox.navigator.RouterRefreshCallback
+import io.mockk.CapturingSlot
+import io.mockk.Runs
+import io.mockk.every
+import io.mockk.just
+import io.mockk.mockk
+import io.mockk.mockkStatic
+import io.mockk.slot
+import io.mockk.unmockkStatic
+import junit.framework.Assert.assertEquals
+import junit.framework.Assert.assertNotNull
+import junit.framework.Assert.assertTrue
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.net.URL
+
+class RouterInterfaceDelegateTest {
+
+    private lateinit var mockRouter: Router
+
+    private companion object {
+        private const val VALID_URL = "https://mapbox.com"
+    }
+
+    @Before
+    fun setUp() {
+        mockRouter = mockk()
+
+        mockkStatic(RouteOptions::class)
+        mockkStatic(DirectionsRoute::class)
+    }
+
+    @After
+    fun cleanUp() {
+        unmockkStatic(RouteOptions::class)
+        unmockkStatic(DirectionsRoute::class)
+    }
+
+    @Test
+    fun routeRequestSuccess() {
+        val routerInterface = provideRouteInterfaceDelegate(mockRouter)
+        val routerCallback = slot<RouterCallback>()
+        val requestId = 5L
+        every { RouteOptions.fromUrl(any()) } returns mockk()
+        every { mockRouter.getRoute(any(), capture(routerCallback)) } returns requestId
+        val (nativeRouterCallback, slotNativeRouterCallback, slotNativeRouterOrigin) =
+            provideNativeRouteCallbackWithSlots()
+        val expectedRoute = provideDirectionsRoute()
+
+        val receivedRequestId = routerInterface.getRoute(VALID_URL, nativeRouterCallback)
+        routerCallback.captured.onRoutesReady(
+            listOf(expectedRoute),
+            RouterOrigin.Onboard,
+        )
+
+        assertNotNull(slotNativeRouterCallback.captured)
+        assertTrue(slotNativeRouterCallback.captured.isValue)
+        val directionsResponse =
+            DirectionsResponse.fromJson(slotNativeRouterCallback.captured.value!!)
+        assertEquals(expectedRoute, directionsResponse.routes().first())
+        assertEquals(com.mapbox.navigator.RouterOrigin.ONBOARD, slotNativeRouterOrigin.captured)
+        assertEquals(requestId, receivedRequestId)
+    }
```

Comment:
Sometimes it's hard to pick a right test double for your specific test. Mocks work better for one tests, stubs work better for other.
I've quickly rewritten this test out of curiosity using stub instead of mock. @RingerJK , please review and give me your feedback: do you find that it's easier to read than with mocks?
```suggestion
    fun routeRequestSuccess() {
        val testRouter = RouterStub()
        val routerInterface = provideRouteInterfaceDelegate(testRouter)
        val expectedRoute = provideDirectionsRoute()

        var receivedResponse: Expected<RouterError, String>? = null
        val receivedRequestId = routerInterface.getRoute(VALID_URL) { response, _ ->
            receivedResponse = response
        }
        testRouter.successfullyCompleteGetRouteRequest(
            receivedRequestId,
            listOf(expectedRoute)
        )

        assertNotNull(receivedResponse)
        assertTrue(receivedResponse!!.isValue)
        val directionsResponse = DirectionsResponse.fromJson(receivedResponse!!.value!!)
        assertEquals(expectedRoute, directionsResponse.routes().first())
    }
```

I will respect your opinion if you just feel that it's easier to read with mocks. But please, give the stub a chase. Try to read it. 🙂 
@mapbox/navigation-android, don't hesitate to express you opinion. You feedback will be very valuable because me and @RingerJK have written one of the options and our views are biased. cc @LukasPaczos , @Zayankovsky .

Here's code of the stub. It's not fully implemented because I'm not sure if you like it:
```kotlin
class RouterStub : Router {

        private var nextRequestId = 0L
        private data class GetRouteRequest(val options: RouteOptions, val callback: RouterCallback)
        private val queuedRouterRequestCallbacks = mutableMapOf<Long, GetRouteRequest>()

        override fun getRoute(routeOptions: RouteOptions, callback: RouterCallback): Long {
            val requestId = nextRequestId++
            queuedRouterRequestCallbacks[requestId] = GetRouteRequest(routeOptions, callback)
            return requestId
        }

        override fun cancelRouteRequest(requestId: Long) {
            queuedRouterRequestCallbacks[requestId]!!.let {
                it.callback.onCanceled(it.options, RouterOrigin.Custom())
            }
            queuedRouterRequestCallbacks.remove(requestId)
        }

        override fun cancelAll() {
            queuedRouterRequestCallbacks.clear()
        }

        override fun getRouteRefresh(
            route: DirectionsRoute,
            legIndex: Int,
            callback: RouteRefreshCallback
        ): Long {
            TODO("Not yet implemented")
        }

        override fun cancelRouteRefreshRequest(requestId: Long) {
            TODO("Not yet implemented")
        }

        override fun shutdown() {
            TODO("Not yet implemented")
        }

        fun successfullyCompleteGetRouteRequest(
            receivedRequestId: Long,
            routes: List<DirectionsRoute>
        ) {
            queuedRouterRequestCallbacks[receivedRequestId]!!.callback.onRoutesReady(
                routes,
                RouterOrigin.Custom()
            )
        }
    }
```